kernel/kernel-api/src/main/java/io/delta/kernel/internal/snapshot/MetadataCleanup.java

```diff
@@ -0,0 +1,198 @@
+   *                   not needed.
+   *               <li>Step 1.2: add `lastCheckpointFiles` to `potentialFileStoDelete` list. This
+   *                   checkpoint is potential candidate to delete later if we find another
+   *                   checkpoint
+   *             </ul>
+   *         <li>Step 2: If the timestamp falls within the retention period, stop
+   *         <li>Step 3: If the file is a delta log file, add it to the `potentialFilesToDelete`
+   *             list
+   *         <li>Step 4: If the file is a checkpoint file, add it to the `lastSeenCheckpointFiles`
+   *       </ul>
+   * </ul>
+   *
+   * @param engine {@link Engine} instance to delete the expired log files
+   * @param clock {@link Clock} instance to get the current time. Useful in testing to mock the
+   *     current time.
+   * @param tablePath Table location
+   * @param retentionMillis Log file retention period in milliseconds
+   * @throws IOException if an error occurs while deleting the log files
+   */
+  public static void cleanupExpiredLogs(
+      Engine engine, Clock clock, Path tablePath, long retentionMillis) throws IOException {
+    checkArgument(retentionMillis >= 0, "Retention period must be non-negative");
+
+    List<String> potentialLogFilesToDelete = new ArrayList<>();
+    long lastSeenCheckpointVersion = -1; // -1 indicates no checkpoint seen yet
+    List<String> lastSeenCheckpointFiles = new ArrayList<>();
+
+    long fileCutOffTime = clock.getTimeMillis() - retentionMillis;
+    logger.info("{}: Starting the deletion of log files older than {}", tablePath, fileCutOffTime);
+    long numDeleted = 0;
+    try (CloseableIterator<FileStatus> files = listDeltaLogs(engine, tablePath)) {
+      while (files.hasNext()) {
+        // Step 1: Check if the `lastSeenCheckpointFiles` contains a complete checkpoint
+        Optional<CheckpointInstance> lastCompleteCheckpoint =
+            getLatestCompleteCheckpointFromList(
+                lastSeenCheckpointFiles.stream().map(CheckpointInstance::new).collect(toList()),
+                CheckpointInstance.MAX_VALUE);
+
+        if (lastCompleteCheckpoint.isPresent()) {
+          // Step 1.1: delete all files in `potentialFilesToDelete`. Now we know there is a
+          //   checkpoint that contains the compacted Delta log up to the checkpoint version and all
+          //   commit/checkpoint files before this checkpoint version are not needed. add
+          //   `lastCheckpointFiles` to `potentialFileStoDelete` list. This checkpoint is potential
+          //   candidate to delete later if we find another checkpoint
+          if (!potentialLogFilesToDelete.isEmpty()) {
+            logger.info(
+                "{}: Deleting log files (start = {}, end = {}) because a checkpoint at "
+                    + "version {} indicates that these log files are no longer needed.",
+                tablePath,
+                getFirst(potentialLogFilesToDelete),
```

Comment:
IMO we should remove these `getFirst` and `getLast` methods. Or, if you want them to stay, can you explain the benefit?

Reply:
Oh, I just meant can we just do `potentialLogFilesToDelete.get(0)` instead of creating a new `getFirst` method.

If you think `getFirst` is cleaner, SGTM.

Reply:
I see. This will be a staple in the later JDK versions, and it is clean rather than trying to say `(list.length - 1)`. It will be useful in many other places.